PAPER.md

```diff
@@ -0,0 +1,378 @@
+
+* `prev hash` field will be the SHA256 sum of the last acknowledged message in the conversation.
+
+* `timestamp` field is an unsigned four-octet number containing the number of seconds elapsed since midnight, 1 January 1970 UTC. Represents the moment of the message signing, which should be close to the send time.
+
+* `source` field is the SHA256 hash value of the sender public key. Creator of the message should add its public key fingerprint here.
+
+* `destination` field is the SHA256 hash value of the receiver public key.
+
+* `headers` is a variable length field. It can be used to add extra metadata to the message in a key value fashion. The specific format for this field could be borrowed from the [HTTP headers standard](https://www.rfc-editor.org/rfc/rfc7230#section-3.2).
+  
+* `data` a variable length field for the content of the message, which will accomplish the [Open PGP message format](https://www.rfc-editor.org/rfc/rfc4880).
+  
+* `signature` its a variable length field for hosting the signature of the whole message. The signature will be computed by putting the hashes of all relevant message fields together, separated by a line break. Here is an example of how the `message data block` would be:
+
+```bash
+uuid
+timestamp
+source
+destination
+SHA256(headers)
+SHA256(data)
+```
+Then the above block would be hashed by `SHA256(message_hashes_block)` . The resultant hash would be signed with the sender private key. Finally, the signature would be appended to the end of the message.
+
+The only encrypted field of the message will be the `data` field, dedicated to the [Open PGP message format](https://www.rfc-editor.org/rfc/rfc4880). Rest of the fields are probably going to be needed for further processing of the messages by other parts of the system, thus they are signed, but not encrypted.
+
+Although [Open PGP message format](https://www.rfc-editor.org/rfc/rfc4880) standard allows adding metadata to their format, its important to mention that many `PGP` implementations are limited in terms the amount of metadata an user can access. So the use of `PGP` on this proposal is limited to its basic functions, and should remain scoped in the `data` field.
+
+### Message verification process
+
+Each peer should always verify all the received messages. Multiple verifications should happen here per each message:
+
+1. The signature verification at DC message level. It should replay the signature hashing process (see Message section) and compare the resultant hash with the signature one, which should be decrypted with the sender public key.
+
+2. The `timestamp` should not be too in the past nor in the future. Of course, we cannot provide accurate measurements here, but should work in a best effort way.
+
+3. The `UUID` should be present in the message header.
```

Comment:
Isn't `uuid` a dedicated field and not a header? Probably missing something here

Reply:
Ok, here it is:
https://github.com/eloylp/decentra-chat/pull/1#discussion_r1010331158

Reply:
No, you are right. A short circuit brain with other bad idea i previously had. Fixed at 9900b4a1480fcb7a9e27dfedb908ef8dcf76488b